### DISCUSSION TITLE
Nocturnal/diurnal filter?

### DISCUSSION THREAD

**Discussion by tphakala** [birdnet-team/BirdNET-Analyzer#250] [General]
How feasible it would be to add time of day filter? I guess large number of species are either diurnal or nocturnal so time of day could be used to filter our false positives. Is such data available at eBird or in any other database?

**Comment by kahst** [comment]
Good question! We've tried this before, but never really followed through. Theoretically, we could try to use the recording times from Xeno-canto to estimate the call times of each bird species. Assuming that the recording times are local time, you could get a pretty accurate picture (with some biases of course, but eBird has those too).

Here is what this would look like for a few selected species (normalized calling probability):

![image](https://github.com/kahst/BirdNET-Analyzer/assets/6673450/b1bfc970-5d9b-43df-aa03-a8131c1ceda1)

It looks like we can correctly derive the diurnal calling patterns for each species. So we could add another filter to sort out owl detections during the day. Maybe a feature request for the near future :) 

In the meantime, if you want to play around with the numbers, you can find the raw data for all species below. Based on the number of recordings per hour, we normalized the call probability between 0 and 100. If there wasn't enough data for a species (the XC metadata is a mess), we used 100 for each hour:

[Download species_recording_times.json here](https://github.com/kahst/BirdNET-Analyzer/files/14207380/species_recording_times.json)

**Reply by tphakala** [reply]
This species recording times data is very useful, I can create a filter of my own based on this. Thank you!

**Reply by kahst** [reply]
Great! Let us know if you find something that could be useful to others and we'll add it to the project.

**Reply by tphakala** [reply]
My implementation will be based on Golang as it will be for my BirdNET-Go app, but I guess I can try to produce Python version also with help of ChatGPT

**Reply by kahst** [reply]
Awesome, thanks!

**Reply by Mattk70** [reply]
Interesting take on using a nocturnal filter! [Chirpity](https://chirpity.mattkirkland.co.uk/) has one. However, the semantics are slightly different. Chirpity's Nocturnal filter excludes all detections in the hours between the start and end of civil twighlight. (These are the "official" nocmig hours). It uses the file's timestamp and the lat / lon of your location to calculate this. 

@tphakala You may well get decent results for the kind of filter you are thinking of without the effort to figure out whether it was winter or summer (and therefore dark or not). All you need to know is whether the species is a nocturnal / diurnal / anytime caller. Looking at @kahst 's charts, it wouldn't be too difficult to infer this from the general pattern of calls. Convert that to a ternary filter and you have what you need - I'd love to know if you crack it!

**Reply by Mattk70** [reply]
Here's a script to do that:

```
import json

x_factor = 2

def analyze_species_calling_patterns(json_data):
    species_data = json.loads(json_data)
    
    night_species = {}
    day_species = {}
    all_hours_species = {}
    night = 0
    day = 0
    all = 0


    for species, calling_hours in species_data.items():
        night_calls = sum(calling_hours[0:6]) + sum(calling_hours[21:])
        day_calls = sum(calling_hours[9:18])
        all_hours_calls = sum(calling_hours)
        
        # 2x more likely to call at night (prevent division by zero)
        if night_calls / (day_calls + 1e-7) > x_factor:
            night+=1
            night_species[species] = 0
            print (species, 'x_factor: ', round(night_calls / (day_calls + 1e-7), 2) )
        elif day_calls / (night_calls + 1e-7) > x_factor:
            day += 1
            day_species[species] = 1
        else:
            all_hours_species[species] = 2
            all += 1
    print('night:', night, 'day: ', day, 'all: ', all)
    result = {**night_species, **day_species, **all_hours_species}
    return result
json_data = <paste json file here between triple quotes>


result = analyze_species_calling_patterns(json_data)

json_result = json.dumps(result)

print(json_result)
```

You'll want to play around with the x_factor and cut-off hours, but the results don't seem too bad?

night:  274 , day:  5641 , all:  607

**Reply by Mattk70** [reply]
I have this working for BirdNET in Chirpity, It's really quite promising! A few notable omissions, but still.

Theses are the results from one 3 hour file, no location filter, just this new nocturnal bird filter:

![image](https://github.com/kahst/BirdNET-Analyzer/assets/61826357/efbe1b71-cb36-4cb0-947d-f74c551186c2)

No, my nights aren't that exciting, but these *are* the right kind of birds to be looking for, for the most part...

**Reply by Mattk70** [reply]
Compare this to the location filter for that week with threshold 0.01:

![image](https://github.com/kahst/BirdNET-Analyzer/assets/61826357/dcf98eb6-2139-48bd-b398-19b6578afc19)
![image](https://github.com/kahst/BirdNET-Analyzer/assets/61826357/246cd29a-d41e-4204-a5ab-8bf6ad42a42d)

**Reply by kahst** [reply]
Cool! So, do you think the metadata model would benefit from a diurnal filter? Or should it be a separate filter?

**Reply by Mattk70** [reply]
It’s a separate filter, I think. You could allow ppl to combine them? i.e location filter and, let’s call it ‘behaviour filter’, so you wouldn’t get nocturnal/diurnal birds reported that wouldn’t be in your location. If you build this kind of filter into the metadata model you wouldn't be able to use it with a custom list. That's if I've understood you correctly?

**Reply by Mattk70** [reply]
I've published a pre-release with this filter enabled here: https://github.com/Mattk70/Chirpity-Electron/releases/tag/v1.1.0. If you haven't installed Chirpity before, do refer to [the website ](https://chirpity.mattkirkland.co.uk) to understand how to navigate the installation process. Also make sure to go to the Settings menu, select Birdnet from the model dropdown, and 'Nocturnal birds" as the list.
Let me know if this new filter is effective or how it might be improved? It's always good to have somehting to kick about to get a better handle on these things!

**Reply by Mattk70** [reply]
Really sorry folks, but I have pulled that pre-release. I can't figure out why, but the build came from the main branch of my repo, not the one in which I'd made the changes. I also can't figure out how to publish from that branch at all.

**Reply by Mattk70** [reply]
I have released [v1.1.0 of Chirpity](https://chirpity.mattkirkland.co.uk), which contains this filter, along with the ability to combine the filter with the 'Local Birds' filter that uses lat/lon and optionally week to refine results.
